tests/unit/model/message_test.py

```diff
@@ -4309,3 +4312,236 @@ def test_boolean_variable_as_condition() -> None:
+        ),
+        (
+            [
+                Link(INITIAL, Field("Tag_1")),
+                Link(
+                    Field("Tag_1"),
+                    Field("Tag_2"),
+                    condition=Equal(Variable("Tag_1"), Variable("TLV::Msg_Data")),
+                ),
+                Link(
+                    Field(ID("Tag_2", location=Location((10, 20)))),
+                    FINAL,
+                    condition=Equal(Variable("Tag_1"), Variable("TLV::Msg_Data")),
+                ),
+            ],
+            {
+                Field("Tag_1"): TLV_TAG,
+                Field("Tag_2"): TLV_TAG,
+            },
+        ),
+    ],
+)
+def test_always_true_message_condition(
+    structure: abc.Sequence[Link], types: abc.Mapping[Field, Type]
+) -> None:
+    link_to_final = [l for l in structure if l.target == FINAL][0]
+    assert_message_model_error(
+        structure,
+        types,
+        (
+            rf'^<stdin>:10:20: model: error: condition "{link_to_final.condition}"'
+            rf' on transition "{link_to_final.source.identifier}" -> "Final" is always true$'
+        ),
+    )
+
+
+def test_possibly_always_true_refinement(
+    monkeypatch: MonkeyPatch, capsys: CaptureFixture[str]
+) -> None:
+    message = Message(
+        "P::M",
+        [
+            Link(INITIAL, Field("Tag")),
+            Link(Field("Tag"), Field("Value")),
+            Link(Field("Value"), FINAL),
+        ],
+        {
+            Field("Tag"): TLV_TAG,
+            Field("Value"): OPAQUE,
+        },
+    )
+    condition = Or(
+        Equal(Variable("Tag"), Variable("TLV::Msg_Data")),
+        Equal(Variable("Tag"), Variable("TLV::Msg_Error")),
+    )
+    monkeypatch.setattr(Proof, "result", ProofResult.UNKNOWN)
+    Refinement(
+        "In_Message",
+        message,
+        Field(ID("Value", location=Location((10, 20)))),
+        MESSAGE,
+        condition,
+    ).error.propagate()
+    captured = capsys.readouterr()
+    assert (
+        f'<stdin>:10:20: model: warning: condition "{condition}"'
+        ' in refinement of "P::M" might be always false'
+    ) in captured.out
+
+
+def test_possibly_always_true_message_condition(
+    monkeypatch: MonkeyPatch, capsys: CaptureFixture[str]
+) -> None:
+    message = Message(
+        "P::M",
+        [
+            Link(INITIAL, Field("Tag")),
+            Link(
+                Field(ID("Tag", location=Location((10, 20)))),
+                FINAL,
+                condition=Equal(Variable("Tag"), Variable("TLV::Msg_Data")),
+            ),
+        ],
+        {
+            Field("Tag"): TLV_TAG,
+        },
+    )
+    monkeypatch.setattr(Proof, "result", ProofResult.UNKNOWN)
+    # pylint: disable=protected-access
+    message._prove_static_conditions()
+    captured = capsys.readouterr()
+    assert (
+        '<stdin>:10:20: model: warning: condition "Tag = TLV::Msg_Data"'
+        ' on transition "Tag" -> "Final" might be always true'
+    ) in captured.out
```

Comment:
Calling a private method to test the functionality is not nice. It should be possible to do the monkey patching before creating the message:

```diff
diff --git a/tests/unit/model/message_test.py b/tests/unit/model/message_test.py
index 6ae1c9806..312264ea9 100644
--- a/tests/unit/model/message_test.py
+++ b/tests/unit/model/message_test.py
@@ -4523,7 +4523,8 @@ def test_possibly_always_true_refinement(
 def test_possibly_always_true_message_condition(
     monkeypatch: MonkeyPatch, capsys: CaptureFixture[str]
 ) -> None:
-    message = Message(
+    monkeypatch.setattr(Proof, "result", ProofResult.UNKNOWN)
+    Message(
         "P::M",
         [
             Link(INITIAL, Field("Tag")),
@@ -4537,9 +4538,6 @@ def test_possibly_always_true_message_condition(
             Field("Tag"): TLV_TAG,
         },
     )
-    monkeypatch.setattr(Proof, "result", ProofResult.UNKNOWN)
-    # pylint: disable=protected-access
-    message._prove_static_conditions()
     captured = capsys.readouterr()
     assert (
         '<stdin>:10:20: model: warning: condition "Tag = TLV::Msg_Data"'
```

The test fails after that change, because the warning is shown twice, which indicates a bug. Could you please investigate what the reason for this is?

Reply:
I tried this. The problem is that the message does many proofs before it reaches the function under test. So the message creation will fail before `_prove_static_conditions` is called. The other alternative would be to monkeypatch `message.verify` too but I wanted to avoid removing all other checks as I want to ensure that this warning is triggered on a valid message.

Reply:
> So the message creation will fail before _prove_static_conditions is called.

I cannot reproduce that. The code shown above only leads to two warnings for me.

Reply:
This is what I get:
```
$ git diff  
diff --git a/tests/unit/model/message_test.py b/tests/unit/model/message_test.py
index 6ae1c980..312264ea 100644
--- a/tests/unit/model/message_test.py
+++ b/tests/unit/model/message_test.py
@@ -4523,7 +4523,8 @@ def test_possibly_always_true_refinement(
 def test_possibly_always_true_message_condition(
     monkeypatch: MonkeyPatch, capsys: CaptureFixture[str]
 ) -> None:
-    message = Message(
+    monkeypatch.setattr(Proof, "result", ProofResult.UNKNOWN)
+    Message(
         "P::M",
         [
             Link(INITIAL, Field("Tag")),
@@ -4537,9 +4538,6 @@ def test_possibly_always_true_message_condition(
             Field("Tag"): TLV_TAG,
         },
     )
-    monkeypatch.setattr(Proof, "result", ProofResult.UNKNOWN)
-    # pylint: disable=protected-access
-    message._prove_static_conditions()
     captured = capsys.readouterr()
     assert (
         '<stdin>:10:20: model: warning: condition "Tag = TLV::Msg_Data"'
$ python -m pytest "tests/unit/model/message_test.py::test_possibly_always_true_message_condition" -vv -s
================================================================================================================================== test session starts ==================================================================================================================================
platform linux -- Python 3.9.9, pytest-7.1.3, pluggy-1.0.0 -- /.../RecordFlux/.venv/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> verbosity=Verbosity.verbose, database=DirectoryBasedExampleDatabase('/.../RecordFlux/.hypothesis/examples')
rootdir: /.../RecordFlux, configfile: pytest.ini
plugins: timeout-2.1.0, xdist-2.5.0, cov-3.0.0, hypothesis-6.23.4, forked-1.4.0
collected 1 item                                                                                                                                                                                                                                                                        

tests/unit/model/message_test.py::test_possibly_always_true_message_condition <stdin>:10:20: model: warning: condition "Tag = TLV::Msg_Data" on transition "Tag" -> "Final" might be always true
<stdin>:10:20: model: warning: condition "Tag = TLV::Msg_Data" on transition "Tag" -> "Final" might be always true
FAILED

======================================================================================================================================= FAILURES ========================================================================================================================================
______________________________________________________________________________________________________________________ test_possibly_always_true_message_condition ______________________________________________________________________________________________________________________
Traceback (most recent call last):
  File "/.../RecordFlux/tests/unit/model/message_test.py", line 4527, in test_possibly_always_true_message_condition
    Message(
  File "/.../RecordFlux/rflx/model/message.py", line 873, in __init__
    self.verify()
  File "/.../RecordFlux/rflx/model/message.py", line 893, in verify
    self._prove_reachability()
  File "/.../RecordFlux/rflx/model/message.py", line 1752, in _prove_reachability
    paths.append((path, proof.error))
  File "/.../RecordFlux/rflx/expression.py", line 74, in error
    assert self._result != ProofResult.SAT
AssertionError
================================================================================================================================ short test summary info ================================================================================================================================
FAILED tests/unit/model/message_test.py::test_possibly_always_true_message_condition - AssertionError
=================================================================================================================================== 1 failed in 5.64s ===================================================================================================================================
```

Reply:
I can reproduce it now, not sure what I did differently before. I still think there is a bug in the handling of the warning message. When all other functions which use `Proof` are disabled, the warning is shown multiple times.

```diff
diff --git a/tests/unit/model/message_test.py b/tests/unit/model/message_test.py
index 6ae1c9806..55384771c 100644
--- a/tests/unit/model/message_test.py
+++ b/tests/unit/model/message_test.py
@@ -4523,7 +4523,14 @@ def test_possibly_always_true_refinement(
 def test_possibly_always_true_message_condition(
     monkeypatch: MonkeyPatch, capsys: CaptureFixture[str]
 ) -> None:
-    message = Message(
+    monkeypatch.setattr(Proof, "result", ProofResult.UNKNOWN)
+    monkeypatch.setattr(Message, "_prove_reachability", lambda x: None)
+    monkeypatch.setattr(Message, "_prove_contradictions", lambda x: None)
+    monkeypatch.setattr(Message, "_prove_coverage", lambda x: None)
+    monkeypatch.setattr(Message, "_prove_overlays", lambda x: None)
+    monkeypatch.setattr(Message, "_prove_field_positions", lambda x: None)
+    monkeypatch.setattr(Message, "_prove_message_size", lambda x: None)
+    Message(
         "P::M",
         [
             Link(INITIAL, Field("Tag")),
@@ -4537,11 +4544,8 @@ def test_possibly_always_true_message_condition(
             Field("Tag"): TLV_TAG,
         },
     )
-    monkeypatch.setattr(Proof, "result", ProofResult.UNKNOWN)
-    # pylint: disable=protected-access
-    message._prove_static_conditions()
     captured = capsys.readouterr()
     assert (
         '<stdin>:10:20: model: warning: condition "Tag = TLV::Msg_Data"'
         ' on transition "Tag" -> "Final" might be always true'
-    ) in captured.out
+    ) == captured.out
```

```console
% python -m pytest -vv tests/unit/model/message_test.py::test_possibly_always_true_message_condition     
=============================================================================================== test session starts ================================================================================================
platform linux -- Python 3.10.8, pytest-7.1.3, pluggy-1.0.0 -- /home/devel/Componolit/RecordFlux/.venv/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> verbosity=Verbosity.verbose, database=DirectoryBasedExampleDatabase('/home/devel/Componolit/RecordFlux/.hypothesis/examples')
rootdir: /home/devel/Componolit/RecordFlux, configfile: pyproject.toml
plugins: forked-1.4.0, cov-3.0.0, xdist-2.5.0, hypothesis-6.23.4, timeout-2.1.0
collected 1 item                                                                                                                                                                                                   

tests/unit/model/message_test.py::test_possibly_always_true_message_condition FAILED                                                                                                                         [100%]

===================================================================================================== FAILURES =====================================================================================================
___________________________________________________________________________________ test_possibly_always_true_message_condition ____________________________________________________________________________________
tests/unit/model/message_test.py:4548: in test_possibly_always_true_message_condition
    assert (
E   assert ('<stdin>:10:20: model: warning: condition "Tag = TLV::Msg_Data" on transition '\n '"Tag" -> "Final" might be always true') == ('<stdin>:10:20: model: warning: condition "Tag = TLV::Msg_Data" on transition '\n '"Tag" -> "Final" might be always true\n'\n '<stdin>:10:20: model: warning: condition "Tag = TLV::Msg_Data" on transition '\n '"Tag" -> "Final" might be always true\n'\n '<stdin>:10:20: model: warning: condition "Tag = TLV::Msg_Data" on transition '\n '"Tag" -> "Final" might be always true\n'\n '<stdin>:10:20: model: warning: condition "Tag = TLV::Msg_Data" on transition '\n '"Tag" -> "Final" might be always true\n')
E     - <stdin>:10:20: model: warning: condition "Tag = TLV::Msg_Data" on transition "Tag" -> "Final" might be always true
E     ?                                                                                                                   -
E     + <stdin>:10:20: model: warning: condition "Tag = TLV::Msg_Data" on transition "Tag" -> "Final" might be always true
E     - <stdin>:10:20: model: warning: condition "Tag = TLV::Msg_Data" on transition "Tag" -> "Final" might be always true
E     - <stdin>:10:20: model: warning: condition "Tag = TLV::Msg_Data" on transition "Tag" -> "Final" might be always true
E     - <stdin>:10:20: model: warning: condition "Tag = TLV::Msg_Data" on transition "Tag" -> "Final" might be always true
============================================================================================= short test summary info ==============================================================================================
FAILED tests/unit/model/message_test.py::test_possibly_always_true_message_condition - assert ('<stdin>:10:20: model: warning: condition "Tag = TLV::Msg_Data" on transition '\n '"Tag" -> "Final" might be alway...
================================================================================================ 1 failed in 5.41s =================================================================================================
```

Reply:
There was a bug that warnings are reported every time `propagate` is called but not cleared (which is not a problem for errors as an exception is raised and `propagate` cannot be called twice).